src/lib/support/OwnerOf.h

```diff
@@ -0,0 +1,31 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <cstddef>
+
+namespace chip {
+
+template<class ClassType, class MemberType>
+constexpr ClassType* OwnerOf(MemberType *ptr, const MemberType ClassType::*member) {
+    return reinterpret_cast<ClassType*>(reinterpret_cast<uintptr_t>(ptr) - offsetof(ClassType, member));
```

Comment:
```suggestion
    static_assert(std::is_standard_layout<ClassType>::value, "offsetof does not make sense");
    return reinterpret_cast<ClassType*>(reinterpret_cast<uintptr_t>(ptr) - offsetof(ClassType, member));
```

Reply:
`is_standard_layout` is too restrictive, and is not necessary.

GCC will raise a compile error when `offsetof` is not applicable. Following code can not compile:

```
#include <cstddef>

struct A { int a; };
struct B : public virtual A   { int b; };

int main() {
    offsetof(A, A::a);
    offsetof(B, B::a);
    return 0;
}
```

```
% gcc -std=c++11 offsetof.cpp -lstdc++
In file included from /usr/include/c++/10/cstddef:50,
                 from offsetof.cpp:1:
offsetof.cpp: In function ‘int main()’:
offsetof.cpp:8:20: error: invalid access to non-static data member ‘A::a’ in virtual base of NULL object
    8 |     offsetof(B, B::a);
      |                    ^
```

Reply:
It's undefined behavior in C++11. Is CHIP using C++17?

Reply:
In practical, we will get an error when it is undefined behavior. And in practical, even it is not standard layer, it is still well defined by implementation (GCC or LLVM). That's why I didn't add the check.

Reply:
> In practical, we will get an error when it is undefined behavior.

What guarantees that?

Why is standard layout too restrictive?